bulk-user-setup/bulk-user-creation.sh

```diff
@@ -0,0 +1,267 @@
+
+  # csvjson for converting CSV -> JSON
+  if ! hash csvjson 2>/dev/null; then
+    echo "${RED}ERROR: CSVJSON tool not found${NORMAL}: try installing ${YELLOW}csvkit${NORMAL}"
+    exit 99
+  fi
+
+}
+
+function verify_json_format_includes_field() {
+  local json=$1
+  local field=$2
+
+  ## verify JSON array is not empty
+  if [ $(echo $json | jq -e '. | length') == 0 ]; then
+    echo "${RED}ERROR: input file conversion produced empty result.${NORMAL} Please check input file format."
+    exit 99
+  fi
+
+  ## verify JSON format by checking JUST THE FIRST ITEM has the required field
+  if [ $(echo $json | jq "first(.[] | has(\"${field}\"))") == false ]; then
+    echo "${RED}ERROR: Field not found in input:${NORMAL} ${field}"
+    exit 99
+  fi
+}
+
+function convert_input_file_to_json() {
+  local file=$1
+
+  verify_csv_tools_are_available
+
+  verify_file_exists ${file}
+
+  # read from CSV by using CSVJSON
+  local raw_csv_as_json=$(csvjson $file)
+
+  # verify JSON format  (ie. check mandatory fields are present)
+  verify_json_format_includes_field "${raw_csv_as_json}" "email"
+  verify_json_format_includes_field "${raw_csv_as_json}" "roles"
+
+  # then reformat JSON using JQ
+  local input_as_json=$(echo $raw_csv_as_json \
+    | jq -r -c 'map({
+        "email": .email,
+        "firstName": .firstName,
+        "lastName": .lastName,
+        "roles": (try(.roles | split("|")) // null)
+      })' )
+
+  echo "$input_as_json"
+}
+
+function process_input_file() {
+  local file=$1
+
+  # convert input file to json
+  json="$(convert_input_file_to_json ${CSV_FILE_PATH})"
+  check_exit_code_for_error $? "$json"
+
+  # write headers to output file
+  echo "#,Email,User JSON,IDAM Status,IDAM Response" >> "$OUTPUT_FILE_PATH"
+
+  # strip JSON into individual compact items then process in a while loop
+  echo $json | jq -r -c '.[]' \
+      |  \
+  ( success_counter=0;fail_counter=0;total_counter=0;
+    while IFS= read -r user; do
+      total_counter=$((total_counter+1))
+      email=$(echo $user | jq --raw-output '.email')
+
+      # make call to IDAM
+      response=$(submit_user_registation "$user")
+      
+      # seperate body and status into an array
+      IFS=$'\n' response_array=($response)
+
+      array_length=${#response_array[@]}
+      if [ $array_length -eq 1 ]; then
+        response_body='' # clear body
+        response_status=${response_array[0]}
+      else
+        response_body=${response_array[0]}
+        response_status=${response_array[${array_length}-1]}
+      fi
+
+      if [ $(( response_status )) -gt 199 ] && [ $(( response_status )) -lt 300 ]; then
+        # SUCCESS:
+        success_counter=$((success_counter+1))
+        echo "${total_counter}, ${email}, ${user}, ${GREEN}${response_status}${NORMAL}, ${response_body}"
+      else
+        # FAIL:
+        fail_counter=$((fail_counter+1))
+        echo "${total_counter}, ${email}, ${user}, ${RED}${response_status}${NORMAL}, ${response_body}"
```

Comment:
Wouldn't it be actually better to output raw data fir the failed calls in a separate file or stream? This could allow us to re-run the script for the failed data only. If we re -run the same script for already-succeeded user registrations, it can be quite troublesome (sending out multiple invite emails or failing at repeated api call for the same user).